File: results/0.25/ELF/elf-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

// ELF Header Constants
#define EI_NIDENT 16
#define ELFMAG0 0x7F
#define ELFMAG1 'E'
#define ELFMAG2 'L'
#define ELFMAG3 'F'

// Parser declarations
static HParser *elf_magic;
static HParser *elf_class;
static HParser *elf_data;
static HParser *elf_version;
static HParser *elf_osabi;
static HParser *elf_abiversion;
static HParser *elf_pad;
static HParser *elf_ident;
static HParser *elf_type;
static HParser *elf_machine;
static HParser *elf_entry;
static HParser *elf_phoff;
static HParser *elf_shoff;
static HParser *elf_flags;
static HParser *elf_ehsize;
static HParser *elf_phentsize;
static HParser *elf_phnum;
static HParser *elf_shentsize;
static HParser *elf_shnum;
static HParser *elf_shstrndx;
static HParser *elf_header;
static HParser *program_header;
static HParser *section_header;
static HParser *elf_file;

void init_parsers(void) {
    // ELF Identification
    elf_magic = h_sequence(h_ch(ELFMAG0), h_ch(ELFMAG1), h_ch(ELFMAG2), h_ch(ELFMAG3), NULL);
    elf_class = h_uint8();  // ELFCLASS32/ELFCLASS64
    elf_data = h_uint8();   // ELFDATA2LSB/ELFDATA2MSB
    elf_version = h_uint8(); // EV_CURRENT
    elf_osabi = h_uint8();
    elf_abiversion = h_uint8();
    elf_pad = h_repeat_n(h_uint8(), 7); // Padding bytes
    
    elf_ident = h_sequence(elf_magic, elf_class, elf_data, elf_version,
                          elf_osabi, elf_abiversion, elf_pad, NULL);

    // ELF Header fields
    elf_type = h_uint16();
    elf_machine = h_uint16();
    elf_version = h_uint32();
    elf_entry = h_uint64();
    elf_phoff = h_uint64();
    elf_shoff = h_uint64();
    elf_flags = h_uint32();
    elf_ehsize = h_uint16();
    elf_phentsize = h_uint16();
    elf_phnum = h_uint16();
    elf_shentsize = h_uint16();
    elf_shnum = h_uint16();
    elf_shstrndx = h_uint16();

    // Complete ELF Header
    elf_header = h_sequence(elf_ident, elf_type, elf_machine, elf_version,
                           elf_entry, elf_phoff, elf_shoff, elf_flags,
                           elf_ehsize, elf_phentsize, elf_phnum,
                           elf_shentsize, elf_shnum, elf_shstrndx, NULL);

    // Program Header
    program_header = h_sequence(
        h_uint32(), // p_type
        h_uint32(), // p_flags
        h_uint64(), // p_offset
        h_uint64(), // p_vaddr
        h_uint64(), // p_paddr
        h_uint64(), // p_filesz
        h_uint64(), // p_memsz
        h_uint64(), // p_align
        NULL
    );

    // Section Header
    section_header = h_sequence(
        h_uint32(), // sh_name
        h_uint32(), // sh_type
        h_uint64(), // sh_flags
        h_uint64(), // sh_addr
        h_uint64(), // sh_offset
        h_uint64(), // sh_size
        h_uint32(), // sh_link
        h_uint32(), // sh_info
        h_uint64(), // sh_addralign
        h_uint64(), // sh_entsize
        NULL
    );

    // Complete ELF File
    elf_file = h_sequence(
        elf_header,
        h_many(program_header),
        h_many(section_header),
        NULL
    );
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <elf_file>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("Failed to open file");
        return 1;
    }

    fseek(fp, 0, SEEK_END);
    size_t size = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    uint8_t *buffer = malloc(size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(fp);
        return 1;
    }

    if (fread(buffer, 1, size, fp) != size) {
        perror("Failed to read file");
        free(buffer);
        fclose(fp);
        return 1;
    }

    init_parsers();

    HParseResult *result = h_parse(elf_file, buffer, size);
    if (!result) {
        fprintf(stderr, "Failed to parse ELF file\n");
        free(buffer);
        fclose(fp);
        return 1;
    }

    // Clean up
    h_parse_result_free(result);
    free(buffer);
    fclose(fp);

    printf("Successfully parsed ELF file\n");
    return 0;
}

File: results/0.25/ELF/elf-claude-3-5-sonnet-20241022.rs
use nom::{
    bytes::complete::{tag, take},
    number::complete::{le_u8, le_u16, le_u32, le_u64, be_u8, be_u16, be_u32, be_u64},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct ElfIdent {
    ei_mag: [u8; 4],
    ei_class: u8,
    ei_data: u8,
    ei_version: u8,
    ei_osabi: u8,
    ei_abiversion: u8,
    ei_pad: [u8; 7],
}

#[derive(Debug)]
struct ElfHeader {
    e_ident: ElfIdent,
    e_type: u16,
    e_machine: u16,
    e_version: u32,
    e_entry: u64,
    e_phoff: u64,
    e_shoff: u64,
    e_flags: u32,
    e_ehsize: u16,
    e_phentsize: u16,
    e_phnum: u16,
    e_shentsize: u16,
    e_shnum: u16,
    e_shstrndx: u16,
}

#[derive(Debug)]
struct ProgramHeader {
    p_type: u32,
    p_flags: u32,
    p_offset: u64,
    p_vaddr: u64,
    p_paddr: u64,
    p_filesz: u64,
    p_memsz: u64,
    p_align: u64,
}

#[derive(Debug)]
struct SectionHeader {
    sh_name: u32,
    sh_type: u32,
    sh_flags: u64,
    sh_addr: u64,
    sh_offset: u64,
    sh_size: u64,
    sh_link: u32,
    sh_info: u32,
    sh_addralign: u64,
    sh_entsize: u64,
}

fn parse_elf_ident(input: &[u8]) -> IResult<&[u8], ElfIdent> {
    let (input, magic) = tag(&[0x7f, 0x45, 0x4c, 0x46])(input)?;
    let (input, ei_class) = le_u8(input)?;
    let (input, ei_data) = le_u8(input)?;
    let (input, ei_version) = le_u8(input)?;
    let (input, ei_osabi) = le_u8(input)?;
    let (input, ei_abiversion) = le_u8(input)?;
    let (input, ei_pad) = take(7usize)(input)?;

    Ok((input, ElfIdent {
        ei_mag: magic.try_into().unwrap(),
        ei_class,
        ei_data,
        ei_version,
        ei_osabi,
        ei_abiversion,
        ei_pad: ei_pad.try_into().unwrap(),
    }))
}

fn parse_elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    let (input, e_ident) = parse_elf_ident(input)?;
    
    let parse_num = match e_ident.ei_data {
        1 => |input| {
            let (input, e_type) = le_u16(input)?;
            let (input, e_machine) = le_u16(input)?;
            let (input, e_version) = le_u32(input)?;
            let (input, e_entry) = le_u64(input)?;
            let (input, e_phoff) = le_u64(input)?;
            let (input, e_shoff) = le_u64(input)?;
            let (input, e_flags) = le_u32(input)?;
            let (input, e_ehsize) = le_u16(input)?;
            let (input, e_phentsize) = le_u16(input)?;
            let (input, e_phnum) = le_u16(input)?;
            let (input, e_shentsize) = le_u16(input)?;
            let (input, e_shnum) = le_u16(input)?;
            let (input, e_shstrndx) = le_u16(input)?;
            Ok((input, (e_type, e_machine, e_version, e_entry, e_phoff, e_shoff, e_flags,
                       e_ehsize, e_phentsize, e_phnum, e_shentsize, e_shnum, e_shstrndx)))
        },
        2 => |input| {
            let (input, e_type) = be_u16(input)?;
            let (input, e_machine) = be_u16(input)?;
            let (input, e_version) = be_u32(input)?;
            let (input, e_entry) = be_u64(input)?;
            let (input, e_phoff) = be_u64(input)?;
            let (input, e_shoff) = be_u64(input)?;
            let (input, e_flags) = be_u32(input)?;
            let (input, e_ehsize) = be_u16(input)?;
            let (input, e_phentsize) = be_u16(input)?;
            let (input, e_phnum) = be_u16(input)?;
            let (input, e_shentsize) = be_u16(input)?;
            let (input, e_shnum) = be_u16(input)?;
            let (input, e_shstrndx) = be_u16(input)?;
            Ok((input, (e_type, e_machine, e_version, e_entry, e_phoff, e_shoff, e_flags,
                       e_ehsize, e_phentsize, e_phnum, e_shentsize, e_shnum, e_shstrndx)))
        },
        _ => panic!("Invalid EI_DATA value"),
    };

    let (input, (e_type, e_machine, e_version, e_entry, e_phoff, e_shoff, e_flags,
                 e_ehsize, e_phentsize, e_phnum, e_shentsize, e_shnum, e_shstrndx)) = parse_num(input)?;

    Ok((input, ElfHeader {
        e_ident,
        e_type,
        e_machine,
        e_version,
        e_entry,
        e_phoff,
        e_shoff,
        e_flags,
        e_ehsize,
        e_phentsize,
        e_phnum,
        e_shentsize,
        e_shnum,
        e_shstrndx,
    }))
}

fn parse_program_header(input: &[u8], ei_data: u8) -> IResult<&[u8], ProgramHeader> {
    let parse_num = match ei_data {
        1 => |input| {
            let (input, p_type) = le_u32(input)?;
            let (input, p_flags) = le_u32(input)?;
            let (input, p_offset) = le_u64(input)?;
            let (input, p_vaddr) = le_u64(input)?;
            let (input, p_paddr) = le_u64(input)?;
            let (input, p_filesz) = le_u64(input)?;
            let (input, p_memsz) = le_u64(input)?;
            let (input, p_align) = le_u64(input)?;
            Ok((input, (p_type, p_flags, p_offset, p_vaddr, p_paddr, p_filesz, p_memsz, p_align)))
        },
        2 => |input| {
            let (input, p_type) = be_u32(input)?;
            let (input, p_flags) = be_u32(input)?;
            let (input, p_offset) = be_u64(input)?;
            let (input, p_vaddr) = be_u64(input)?;
            let (input, p_paddr) = be_u64(input)?;
            let (input, p_filesz) = be_u64(input)?;
            let (input, p_memsz) = be_u64(input)?;
            let (input, p_align) = be_u64(input)?;
            Ok((input, (p_type, p_flags, p_offset, p_vaddr, p_paddr, p_filesz, p_memsz, p_align)))
        },
        _ => panic!("Invalid EI_DATA value"),
    };

    let (input, (p_type, p_flags, p_offset, p_vaddr, p_paddr, p_filesz, p_memsz, p_align)) = parse_num(input)?;

    Ok((input, ProgramHeader {
        p_type,
        p_flags,
        p_offset,
        p_vaddr,
        p_paddr,
        p_filesz,
        p_memsz,
        p_align,
    }))
}

fn parse_section_header(input: &[u8], ei_data: u8) -> IResult<&[u8], SectionHeader> {
    let parse_num = match ei_data {
        1 => |input| {
            let (input, sh_name) = le_u32(input)?;
            let (input, sh_type) = le_u32(input)?;
            let (input, sh_flags) = le_u64(input)?;
            let (input, sh_addr) = le_u64(input)?;
            let (input, sh_offset) = le_u64(input)?;
            let (input, sh_size) = le_u64(input)?;
            let (input, sh_link) = le_u32(input)?;
            let (input, sh_info) = le_u32(input)?;
            let (input, sh_addralign) = le_u64(input)?;
            let (input, sh_entsize) = le_u64(input)?;
            Ok((input, (sh_name, sh_type, sh_flags, sh_addr, sh_offset, sh_size,
                       sh_link, sh_info, sh_addralign, sh_entsize)))
        },
        2 => |input| {
            let (input, sh_name) = be_u32(input)?;
            let (input, sh_type) = be_u32(input)?;
            let (input, sh_flags) = be_u64(input)?;
            let (input, sh_addr) = be_u64(input)?;
            let (input, sh_offset) = be_u64(input)?;
            let (input, sh_size) = be_u64(input)?;
            let (input, sh_link) = be_u32(input)?;
            let (input, sh_info) = be_u32(input)?;
            let (input, sh_addralign) = be_u64(input)?;
            let (input, sh_entsize) = be_u64(input)?;
            Ok((input, (sh_name, sh_type, sh_flags, sh_addr, sh_offset, sh_size,
                       sh_link, sh_info, sh_addralign, sh_entsize)))
        },
        _ => panic!("Invalid EI_DATA value"),
    };

    let (input, (sh_name, sh_type, sh_flags, sh_addr, sh_offset, sh_size,
                 sh_link, sh_info, sh_addralign, sh_entsize)) = parse_num(input)?;

    Ok((input, SectionHeader {
        sh_name,
        sh_type,
        sh_flags,
        sh_addr,
        sh_offset,
        sh_size,
        sh_link,
        sh_info,
        sh_addralign,
        sh_entsize,
    }))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <elf-file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1]).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_elf_header(&buffer) {
        Ok((remaining, header)) => {
            println!("ELF Header: {:?}", header);

            // Parse program headers
            let mut current_pos = header.e_phoff as usize;
            for _ in 0..header.e_phnum {
                if let Ok((_, ph)) = parse_program_header(&buffer[current_pos..], header.e_ident.ei_data) {
                    println!("Program Header: {:?}", ph);
                    current_pos += header.e_phentsize as usize;
                }
            }

            // Parse section headers
            let mut current_pos = header.e_shoff as usize;
            for _ in 0..header.e_shnum {
                if let Ok((_, sh)) = parse_section_header(&buffer[current_pos..], header.e_ident.ei_data) {
                    println!("Section Header: {:?}", sh);
                    current_pos += header.e_shentsize as usize;
                }
            }
        }
        Err(e) => eprintln!("Failed to parse ELF header: {:?}", e),
    }
}